### ISSUE TITLE
"rclone merge" - reconcilation and two-way syncing

### ISSUE THREAD

**Issue by lpan** [rclone/rclone#2870]
#### What is your current rclone version (output from `rclone version`)?
```
rclone v1.45
- os/arch: linux/amd64
- go version: go1.11.2
```

#### What problem are you are trying to solve?
Implement a new primitive/sub-command `rclone merge`. We can then use it implement "two-way syncing" or "smart deletion synchronization" as described in https://github.com/ncw/rclone/issues/118

#### How do you think rclone should be changed to solve that?
When data is replicated to two or more places (local file system and Dropbox), clients will be able to apply changes to them independently. Then, the replicated copies of the data will be out-of-sync and potentially conflicted.

https://github.com/ncw/rclone/issues/118 brought up that  `rclone` doesn’t support “smart two-way deletion synchronization”, or just two-way syncing. In my opinion, this is because `rclone` doesn’t support a primitive that provides the capability to merge two different (possibly conflicted) replicated copies of the synced data.

I propose to add such primitive by implementing a sub-command called `rclone merge`.

In short, `rclone merge` will reconcile two out-of-sync directories located in the two file systems (local and remote) using a pluggable conflict resolution strategy (see linked proposal below for more details). Then, it will write the reconciled result locally, and use rclone sync to push the reconciled version of the data to the remote.

If you want to learn more about what I am trying to do, please read this [detailed proposal](https://paper.dropbox.com/doc/Proposal-rclone-merge--AUhse6SUm8I4elPZS5AX_dh8AQ-uoRilZxIqDhUK24hVY3Ee) hosted on Dropbox Paper. Feel free to comment or contact me via email (lawrencefeipan@gmail.com) if you have any questions. Thanks!

**Comment by ncw** [comment]
Very interesting idea - I'd love to have your help with implementing this for rclone :-)

You might want to examine https://github.com/DavideRossi/upback also and possibly collaborate with the author?

When doing a sync rclone will be matching on (size, modtime) or (size, checksum) or (size) or some other combinations depending on flags!  I'd assume a merge algorithm would probably continue to use the sync routines, but whenever a mismatch was detected, so fileA != fileB (using the above criteria) or FileA present and FileB not present go into the merge algorithm to work out what to do (copy A->B, copy B->A, delete A, delete B)

In your paper you talk about a "file system change log". Alas there is no such thing - at least universally everywhere.  So I think any timestamps you want will have to be synthesized by the merge program.  Perhaps by storing metadata at the time of a sync...

If you wanted to do some more research, you could do worse than finding out how [unison](https://www.cis.upenn.edu/~bcpierce/unison/) works.  There are some [research papers on file synchronization](http://www.cis.upenn.edu/~bcpierce/papers/index.shtml#File%20Synchronization) linked there too

**Comment by lpan** [comment]
Thanks so much Nick :slightly_smiling_face: I am interested in helping out!

> You might want to examine https://github.com/DavideRossi/upback also and possibly collaborate with the author?

Will read up on `upback` for sure :+1: It seems to be something very similar to what I want to accomplish, but as a Python side script of `rclone`.

> When doing a sync rclone will be matching on (size, modtime) or (size, checksum) or (size) or some other combinations depending on flags! I'd assume a merge algorithm would probably continue to use the sync routines, but whenever a mismatch was detected, so fileA != fileB (using the above criteria) or FileA present and FileB not present go into the merge algorithm to work out what to do (copy A->B, copy B->A, delete A, delete B)

Thanks for the suggestion :+1: I will spend some time to dig into the source code later today to find out how `rclone sync` works exactly. Then, I will come up with a more detailed implementation plan for `rclone merge`.

Just a side note, I think it is also important to **decouple** the reconciliation (conflict resolution) strategies from the `rclone merge` command. Smart reconciliation strategies are hard to implement and may require states (rclone metadata). But there also dumb but also easy-to-implement strategies (manual resolution).

> If you wanted to do some more research, you could do worse than finding out how unison works. There are some research papers on file synchronization linked there too

I will definitely read up on Unison in the meanwhile :slightly_smiling_face:

**Comment by ncw** [comment]
> I am feeling a bit burnt out lately. Will re-visit this later  If anyone is interested feel free to take over

No worries!

**Comment by bitcoinmeetups** [comment]
I think rclone merge is a very good idea and very useful. After you have implemented it you could also consider using a flag to activate it instead. It would feel more logical to input rclone sync --two-way instead of rclone merge (because rclone sync defaults to one-way syncing. Which, in a way, isn't really synchronizing at all).

**Comment by lpan** [comment]
I am back :slightly_smiling_face: Starting to look into this again

**Comment by cjnaz** [comment]
@lpan , please also see [rclonesync](https://github.com/cjnaz/rclonesync-V2) for another implementation, and a discussion of the corner cases.  One consideration is fail-safe - ensuring that the tool cant unintentionally delete everything, due to network access problems, for example.  There also needs to be a history mechanism so that changes can be cleanly detected on both paths.

**Comment by lpan** [comment]
Two years later...

I want to give this another shot. You can track my progress here https://paper.dropbox.com/doc/wip-Revisiting-rclone-merge--BBqOGNQBk8ezq7gMVI2vTHh2AQ-CRbnPTJmYHQcn3mIkIh7R

cc @ivandeex

**Comment by cjnaz** [comment]
@lpan please see https://github.com/rclone/rclone/issues/2870#issuecomment-558325865 for design requirements and/or considerations.

**Comment by cjnaz** [comment]
Feedback on your scenarios Dropbox doc...  Please see the [rclonesync](https://github.com/cjnaz/rclonesync-V2) README.md for a discussion of scenarios and corner cases.

**Comment by lpan** [comment]
Listing some alternatives to the feature proposed in this issue. I hope they will help justify why it is important to implement two-way syncing directly in `rclone`.

## 1. [rclone-sync](https://github.com/cjnaz/rclonesync-V2)
It is a Python wrapper on top of rclone that achieves exactly this feature. It works well but has some limitations
* Users need to download another binary + read another wiki to get it to work (steeper setup curve)
* The command-line interface and args are not coherent with `rclone`'s (steeper setup curve)
* [Performance issues](https://github.com/cjnaz/rclonesync-V2/issues/36) (can only issue one rclone command at a time)
* Not being a `rclone` command makes it difficult for users to discover this project

## 2. [syncthing](https://github.com/syncthing/syncthing)
* `syncthing` won't work with dumb storage backends (you have to run another `syncthing` instance on the remote)

## 3. [`rclone mount`](https://rclone.org/commands/rclone_mount/#limitations)
* Too many dependencies (e.g. fuse) + lots of [limitations](https://rclone.org/commands/rclone_mount/#limitations)

**Comment by ivandeex** [comment]
I am going to implement @cjnaz 's `rclonesync` as `rclone bisync` command mentioned earlier.

**Comment by lpan** [comment]
@ivandeex That's amazing!! I will close this issue in favour of #118
